alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,263 @@
 import Cocoa
+import Foundation
 
-class PreferencesPanel: NSPanel, NSTextViewDelegate {
-    var maxScreenUsage: NSTextView?
-    var windowPadding: NSTextView?
-    var cellPadding: NSTextView?
-    var cellBorderWidth: NSTextView?
-    var maxThumbnailsPerRow: NSTextView?
-    var iconSize: NSTextView?
-    var fontHeight: NSTextView?
-    var interItemPadding: NSTextView?
-    var tabKeyCode: NSTextView?
-    var windowDisplayDelay: NSTextView?
-    var metaKey: NSPopUpButton?
-    var theme: NSPopUpButton?
-    var showOnScreen: NSPopUpButton?
-    var inputsMap = [NSTextView: String]()
+class PreferencesPanel: NSPanel, NSWindowDelegate {
+    let panelWidth = CGFloat(496)
+    let panelHeight = CGFloat(256) // auto expands to content height (but does not auto shrink)
+    let panelPadding = CGFloat(40)
+    var labelWidth: CGFloat { get { // derived convenience variable
+        return (panelWidth - panelPadding) * CGFloat(0.45)
+    }}
+    var windowCloseRequested = false
 
     override init(contentRect: NSRect, styleMask style: StyleMask, backing backingStoreType: BackingStoreType, defer flag: Bool) {
-        super.init(contentRect: contentRect, styleMask: style, backing: backingStoreType, defer: flag)
+        let initialRect = NSRect(x: 0, y: 0, width: panelWidth, height: panelHeight)
+        super.init(contentRect: initialRect, styleMask: style, backing: backingStoreType, defer: flag)
         title = Application.name + " Preferences"
         hidesOnDeactivate = false
-        contentView = makeGridView(makeLabelsAndInputs(), makeWarningLabel())
+        contentView = makeContentView()
+        NSApp.activate(ignoringOtherApps: true)
     }
 
-    private func makeLabelsAndInputs() -> [[NSView]] {
+    override func close() {
+        NSApp.deactivate()
+        (NSApp as! Application).preferencesPanel = nil
+        super.close()
+    }
+
+    public func windowShouldClose(_ sender: NSWindow) -> Bool {
+        windowCloseRequested = true
+        challengeNextInvalidEditableTextField()
+        return attachedSheet == nil // user is challenged
+    }
+
+    private func challengeNextInvalidEditableTextField() {
+        let invalidFields = (contentView?
+                .findNestedViews(subclassOf: PreferencesPanelNSTextFieldEditable.self)
+                .filter({ !$0.isValid() })
+        )
+        let focusedField = invalidFields?.filter({ $0.currentEditor() != nil }).first
+        let fieldToNotify = focusedField ?? invalidFields?.first
+        fieldToNotify?.delegate?.controlTextDidChange?(Notification(name: NSControl.textDidChangeNotification, object: fieldToNotify))
+
+        if fieldToNotify != focusedField {
+            makeFirstResponder(fieldToNotify)
+        }
+    }
```

Comment:
Ok I understand now what you do. You block the user from leaving the window if a field is invalid. I think you made it was harder for yourself!

* Easiest/strongest fix: remove these annoying text fields and put sliders. Done. User can make mistakes now
* Intermediate fix: keep the previous logic to reset the field as soon as an incorrect value is written and focus on that field goes gone. This guarantees the field has a correct value always.
* Too-intense fix: the approach you took here where you want to communicate more to the user but I believe you end up confusing them more than previous 2 approaches as the red field doesn't explain what's wrong next to it, and when I close the panel, it says it's wrong but not why, or which format is expected. Also what if I don't close the window but close the app, or simply switch the window to the background and expect my change to be active? I think you're going into a territory that would require integrating a full-blown validation framework, but in our case it's way simpler to do the solutions above.

Reply:
Gotcha. Removed the text fields for the other fields as mentioned in another code review comment. Tab key input is the only one remaining for the moment.

Thats focus/unfocus is my problem. I have issues to get notified when the input looses focus anyhow (panel closed, dropdown opened, another input selected) to trigger a save Preference operation. If we don't trigger the save the last somehow here the value the user entered will not be saved at all. The panel before this pull-request has the same issue. If the user reopened the panel, he saw whatever he entered, even if it was not saved or the Preferences rejected the value. The PreferencesPanel did only save if the user focused an other input. This is my attempt at fixing this without requiring the user to click something.

One of the workarounds I came up with was the invisibleTextField (to trigger an unfocus of the current input which then triggered a save). Now this invisibleTextField workaround is removed and the inputs are triggering a save on every key press. But we need to allow the inputs to be invalid while the user is editing them (or has the panel not closed yet), that means to skip execution of an save (or there will be an exception triggering the error sheet panel). 

One earlier iteration I had it that way, a user was not able to leave an input while it was not valid but that felt not right. Since there is now only one input text field remaining I could try to find out how I did it and reimplement it. 

Once we have some way of recording a key or key combination (#68) we could remove all the TextField handling/complications.

> dosn't explain what's wrong next to it, and when I close the panel, it says it's wrong but not why, or which format is expected. 

Fair enough.

> Also what if I don't close the window but close the app, or simply switch the window to the background and expect my change to be active? 

I handled that 😬  The field validates instantly and gives feedback in time. The Preference will be saved to Preferences and disk if its valid. If the user close forces the window or the application and reopens, he will see his last valid value for the said Preference.

> I think you're going into a territory that would require integrating a full-blown validation framework, but in our case it's way simpler to do the solutions above.

Yes, once the textfields are gone completely this will be a happy little code to look at.